File: src/contexts/chat/transport/rest/dtos/chat-completion-response.dto.ts
import { CreateChatCompletionResult } from '@contexts/chat/application/commands/create-chat-completion/create-chat-completion.handler';
import { ApiProperty } from '@nestjs/swagger';

import { ChatCompletionChoiceDto } from './chat-completion-choice.dto';

/**
 * Mirrors OpenAI's response shape.
 */
export class ChatCompletionResponseDto {
  @ApiProperty()
  id!: string;

  @ApiProperty({ example: 'chat.completion' })
  object!: string;

  @ApiProperty()
  created!: number;

  @ApiProperty()
  model!: string;

  @ApiProperty({ type: [ChatCompletionChoiceDto] })
  choices!: ChatCompletionChoiceDto[];

  static from(
    result: CreateChatCompletionResult,
    requestedModel: string,
  ): ChatCompletionResponseDto {
    const dto = new ChatCompletionResponseDto();
    dto.id = `chatcmpl-${result.jobId}`;
    dto.object = 'chat.completion';
    dto.created = Math.floor(Date.now() / 1000);
    dto.model = requestedModel;
    dto.choices = [
      {
        index: 0,
        message: {
          role: result.result.message.role.value,
          content: result.result.message.content.value,
        },
        finish_reason: result.result.finishReason.value,
      },
    ];
    return dto;
  }
}

File: src/main/java/com/sisqueslabs/nexora/api/contexts/chat/transport/rest/dto/ChatCompletionResponseDto.java
package com.sisqueslabs.nexora.api.contexts.chat.transport.rest.dto;

import java.time.Instant;
import java.util.List;

import com.sisqueslabs.nexora.api.contexts.chat.application.command.createchatcompletion.CreateChatCompletionResult;
import com.sisqueslabs.nexora.api.contexts.chat.domain.valueobjects.MessageValueObject;

/**
 * Mirrors OpenAI's response shape.
 */
public record ChatCompletionResponseDto(
        String id,
        String object,
        long created,
        String model,
        List<ChatCompletionChoiceDto> choices) {

    public static ChatCompletionResponseDto from(CreateChatCompletionResult result, String requestedModel) {
        MessageValueObject message = result.result().message();
        ChatCompletionChoiceDto choice = new ChatCompletionChoiceDto(
                0,
                new ChatCompletionMessageDto(message.role().wireValue(), message.content()),
                result.result().finishReason().wireValue());

        return new ChatCompletionResponseDto(
                "chatcmpl-" + result.jobId(),
                "chat.completion",
                Instant.now().getEpochSecond(),
                requestedModel,
                List.of(choice));
    }
}
